### PR TITLE
Add fastify addons

### LINKED ISSUE #21: cleanup db properly
review data that remains in the db and should no longer be there:

- approved requests
- deleted tokens
- ...

### LINKED ISSUE #33: Use fastify-jwt for auth
The backend uses firebase sdk to verify the auth token. Use fastify-jwt instead.

### PR DESCRIPTION
**Pull request by sameer-coder**
This PR includes:

- Added Fastify sensible
- Added Fastify helmet
- Minor refactor in error handling
- Fastify-jwt used for token verification
- Requests are now deleted after otp check

Closes #21 
Closes #33

### CODE REVIEW

**Review by simoneb — APPROVED** [review]
code wise LGTM, make sure everything keeps working (both web and mobile apps) specifically around the restrictions introduced by helmet

**Comment by simoneb** [comment]
@sameer-coder it would make sense to include #33 in this PR

File: server/lib/plugins/push.js [hunk 1]
```diff
@@ -63,22 +58,11 @@ async function pushPlugin(server, options) {
...
-            secretId,
-            uniqueId
-          })
-        })
-      )
+    send: async ({ subscription, secretId, uniqueId }) => {
+      const params = { subscription, secretId, uniqueId }
+      subscription.type === 'expo'
+        ? sendExpoPush(log, expo, params)
+        : sendWebPush(log, params)
```
Comment (simoneb) [inline server/lib/plugins/push.js:65]: why this change?

File: server/lib/routes/otp.js [hunk 2]
```diff
@@ -39,26 +39,27 @@ async function otpRoutes(server) {
...
       const uniqueId = uniqid()
 
       return new Promise(() => {
-        const completeRequest = (error, otp) => {
+        const completeRequest = (error, otp, requestObj) => {
+          requestObj.delete()
           if (error) {
             log.error(error.message)
-            return reply.code(500).send(error.message)
+            return reply.internalServerError('An unexpected error occured')
```
Comment (simoneb) [inline server/lib/routes/otp.js:54]: for 500s we can omit the error message altogether

File: server/lib/routes/otp.js [hunk 3]
```diff
@@ -39,26 +39,27 @@ async function otpRoutes(server) {
...
-        return reply.code(404).send(`No subscription found for user ${userId}`)
+        return reply.notFound(`No subscription found for user ${userId}`)
       }
 
       const uniqueId = uniqid()
 
       return new Promise(() => {
-        const completeRequest = (error, otp) => {
+        const completeRequest = (error, otp, requestObj) => {
+          requestObj.delete()
```
Comment (simoneb) [inline server/lib/routes/otp.js:51]: isn't this an async operation that we may want to wait?

File: server/lib/routes/otp.js [hunk 4]
```diff
@@ -75,23 +76,27 @@ async function otpRoutes(server) {
...
             log.error(error.message)
-            return reply.code(500).send(error.message)
+            return reply.internalServerError('An unexpected error occured')
           }
         )
 
-        // fire the notification to all available subscription
         push.send({
-          subscriptions: subscriptions.docs,
+          subscription: subscription.docs[0].data(),
```
Comment (simoneb) [inline server/lib/routes/otp.js:88]: I'm still somewhat surprised that we're using a query which returns possibly multiple docs when all we need is to look up 1 doc by its primary key

File: server/lib/routes/otp.js [hunk 5]
```diff
@@ -105,14 +110,14 @@ async function otpRoutes(server) {
...
-        return reply.code(404).send()
+        return reply.notFound('Request does not exist')
       }
 
       await requestObj.update({ otp: otp || null, approved })
       reply.code(201).send()
     } catch (error) {
       request.log.error(error.message)
-      return reply.code(500).send()
+      return reply.internalServerError('An unexpected error occured')
```
Comment (simoneb) [inline server/lib/routes/otp.js:120]: no need for a custom error message

File: server/lib/routes/otp.js (hunk 4 above)
Comment (simoneb) [reply server/lib/routes/otp.js:88]: indeed there is a way to do that https://github.com/nearform/optic/pull/46/files#diff-32a64babda63d3014c589afe5b130e3e927456f69b66a397a0f30715d570eb1aR110

File: server/lib/routes/otp.js [hunk 6]
```diff
@@ -75,23 +76,27 @@ async function otpRoutes(server) {
...
         })
 
-        const timeout = setTimeout(completeRequest, approvalLimit)
+        const timeout = setTimeout(
+          completeRequest,
+          approvalLimit,
+          null,
+          null,
+          requestObj
+        )
```
Comment (simoneb) [inline server/lib/routes/otp.js:99]: not too fond of this new logic. can we simplify this?

**Review by simoneb — COMMENTED** [review]
Overall looks good, see inline comments. Apologies as I requested it but using fastify-jwt doesn't look like a great idea. I actually thought we could use it but still use firebase/auth to verify the token. I think we should keep firebase and not use fastify-jwt instead, as it seems to involve additional, unwanted complexity. I hope it's easy to revert that part alone.

File: server/lib/routes/otp.js (hunk 3 above)
Comment (sameer-coder) [reply server/lib/routes/otp.js:51]: It is an async operation but I don't think we have to necessarily wait

File: server/lib/plugins/push.js (hunk 1 above)
Comment (sameer-coder) [reply server/lib/plugins/push.js:65]: Earlier this method received multiple subscriptions which is why it used `map`. Now this method is called with only one subscription so I have simplified the code

File: server/lib/routes/otp.js (hunk 4 above)
Comment (sameer-coder) [reply server/lib/routes/otp.js]: I have corrected this now

File: server/lib/routes/token.js [hunk 7]
```diff
@@ -25,14 +25,19 @@ async function tokenRoutes(server) {
...
+        await db
+          .collection('tokens')
+          .doc(secretId)
+          .set({
+            token,
+            subscriptionId
+          })
+        reply.send({ token })
+      } catch (error) {
+        request.log.error(error.message)
```
Comment (simoneb) [inline server/lib/routes/token.js:38]: with fastify I think we can log the error directly, can you check what the api for that looks like?

File: server/lib/routes/token.js [hunk 8]
```diff
@@ -46,37 +51,42 @@ async function tokenRoutes(server) {
...
+        }
+
+        await db
+          .collection('tokens')
+          .doc(secretId)
+          .delete()
+
+        reply.code(204).send()
+      } catch (error) {
+        request.log.error(error.message)
```
Comment (simoneb) [inline server/lib/routes/token.js:87]: same as earlier comment

**Review by simoneb — APPROVED** [review]
Just a couple of minor comments but looks good otherwise 🚀

File: server/lib/routes/token.js (hunk 7 above)
Comment (sameer-coder) [reply server/lib/routes/token.js:38]: @simoneb Do you mean allowing the default fastify error handler to handle this? I guess I can get rid of the try-catch since fastify automatically logs the error and sends a 500 reponse.

Comment (simoneb) [reply server/lib/routes/token.js:38]: that's also an option, yes, in fact I think we may need to do a try/catch here at all!
